File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
#include <string>
#include <vector>
#include <algorithm>

class Solution {
public:
    int lengthOfLongestSubstring(std::string s) {
        // Use a vector of size 128 for standard ASCII, initialized to -1
        std::vector<int> charIndex(128, -1);
        
        int left = 0;
        int maxLength = 0;

        // Traverse the string, expanding the window with the right pointer
        for (int right = 0; right < s.length(); ++right) {
            char c = s[right];
            
            // Check if the current character's last seen position is within the active window
            if (charIndex[c] >= left) {
                // Advance the left pointer to immediately resolve the repetition
                left = charIndex[c] + 1;
            }
            
            // Update the dictionary/array with the current character's new index
            charIndex[c] = right;
            
            // Evaluate if the current valid substring is the longest we've seen
            maxLength = std::max(maxLength, right - left + 1);
        }
        
        return maxLength;
    }
};

File: LeetCode/Medium/0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Dictionary to store the most recent index of each character
        char_map = {}
        left = 0
        max_len = 0

        # Iterate through the string with the right pointer
        for right in range(len(s)):
            current_char = s[right]
            
            # If the character is a duplicate AND its last seen index is within the current window
            if current_char in char_map and char_map[current_char] >= left:
                # Instantly move the left pointer to skip the duplicate character
                left = char_map[current_char] + 1
            
            # Update the character's most recent index in the map
            char_map[current_char] = right
            
            # Calculate the current window length and update the maximum length found
            max_len = max(max_len, right - left + 1)

        return max_len